File: sabian_native_common_ios/ios/Classes/channels/methods/AbstractMethodChannelHandler.swift
//
//  AbstractMethodChannelHandler.swift
//  Runner
//
//  Created by bryosabian on 23/11/2023.
//

import Foundation
import Flutter

class AbstractMethodChannelHandler : IChannelHandler<IMethodChannelHandler, MethodChannelPayload> {
    
    override func execute(_ action: IMethodChannelHandler, _ args: MethodChannelPayload) {
        action.execute(payload: args)
    }
    
    override func create(_ key: String, payload : ChannelPayload) {
        
        super.create(key, payload: payload)
        
        let channel = FlutterMethodChannel(name: key, binaryMessenger: payload.messenger)
        
        channel.setMethodCallHandler({ [unowned self] call,result in
            self.execute(channelAction : call.method, args: MethodChannelPayload(controller: controller, result: result, call: call))
        })

    }
}

File: sabian_native_common_android/android/src/main/kotlin/com/sabiantech/sabian_native_common_android/channels/methods/AbstractMethodChannelHandler.kt
package com.sabiantech.sabian_native_common_android.channels.methods

import android.util.Log
import com.sabiantech.sabian_native_common_android.channels.ChannelPayload
import com.sabiantech.sabian_native_common_android.channels.IChannelHandler
import io.flutter.plugin.common.MethodChannel

abstract class AbstractMethodChannelHandler : IChannelHandler<IMethodChannelHandler, MethodChannelPayload>() {

    override fun create(key: String, payload: ChannelPayload) {
        super.create(key, payload)
        Log.e("sabian_common_native", "Created method channels for $key")
        val channel = MethodChannel(payload.messenger, key)
        channel.setMethodCallHandler { call, result ->
            execute(call.method, MethodChannelPayload(payload.context, result, call))
        }
    }

    override fun execute(action: IMethodChannelHandler, args: MethodChannelPayload) {
        action.execute(args)
    }
}